### DISCUSSION TITLE
New Linux user having a hard time (trying to replicate One Drive behavior on Windows)

### DISCUSSION THREAD

**Discussion by reouvenzana** [abraunegg/onedrive#2311] [Q&A] [answered]
Hello,

First, thanks to @abraunegg and all the contributors;
Second, I'm one week only into Linux (Ubuntu 22.04 LTS) and I'm not particularly versed into the technical stuff.

I'm trying to replicate the behavior of OneDrive on Windows, and the documentation has let me a little bit confused. 

Since 'File on Demand' is not possible, I plan to sync only some directories and to use the web app for the rest. 

In Windows, the changes on my files on ComputerA are reflected on the Drive (changes are uploaded). If I modify those files through ComputerB, these changes will also be reflected when I log into ComputerA (changes are downloaded). I don't really understand how to replicate this bi-directional syncing with this extension

(1) How to replicate this behaviour manually? Should I use `onedrive --synchronize --download-only --single-directory <dir-name>` to get the last version of the files on OneDrive and `onedrive --synchronize --upload-only --single-directory <dir-name>` to then upload my modifications? 

(2) I've seen about [running OneDrive as a service ](https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#running-onedrive-as-a-system-service), but I don' understand what it does. Maybe that's what I'm searching for?

Thank you.

**Comment by abraunegg** [comment, marked as answer]
@reouvenzana 
Welcome to the wonderful world of open source!

> I don't really understand how to replicate this bi-directional syncing with this extension

Essentially, when you are using `--download-only` or `--upload-only` you are performing a one-way sync in that direction.

To replicate what you are seeing in Windows, omit those and changes locally & remote will sync. This is what you are looking for.

When using `--single-directory ` - this means you *only* want to sync that directory under your `~/OneDrive` folder .... all other changes will be discarded / not synced.

So if you are trying to just sync your OneDrive account online to --> `~/OneDrive` locally, all you need to run is:

```
onedrive --synchronize
```
or with some extra logging output
```
onedrive --synchronize --verbose
```

To run a a service - essentially it is background syncing all the time - so rather than you running a sync manually - its being done for you - which further replicates what would be occurring on Windows.

Hope this explains what you are looking for.

**Reply by reouvenzana** [reply]
Thank you for the quick answer. Though, I'm still not confident about how to proceed.

So, here is what I've tried so far. As I said, I want to sync one or two directories only and I can handle the rest with the web app (I saw onedriver but that's too much for me now). 

I used `onedrive --synchronize --single-directory <Notes>` to get my notes locally. I've made some changes. Those changes haven't been reflected on OneDrive. So, what's next? If I run the synchronize command once more, what would happen? Will OneDrive files be downloaded again and erase my modifications, or the contrary (my changes will be uploaded to OneDrive)? 

I'll eventually try 'run as a service' but I'll like to take the slow path.

**Reply by abraunegg** [reply]
@reouvenzana 
> I'll eventually try 'run as a service' but I'll like to take the slow path

Add `--verbose --dry-run` to your command

This will show you what would be done, without the changes actually occuring.

Once confident, remove dry-run to sync those changes

As for what changes go where .. it all depends on who has the latest copy of the document based on a number of factors - and if there is a local conflict, the local is renamed so there is no data loss

**Reply by reouvenzana** [reply]
Thank you! I understand better how it works after some tests.

**Comment by abraunegg** [comment]
@reouvenzana 
> Since 'File on Demand' is not possible, I plan to sync only some directories and to use the web app for the rest.

It is available - just not through this application. You need to use https://github.com/jstaf/onedriver instead to use Windows 'on-demand'  functionality on Linux

**Comment by stevefxp1** [comment]
I am going to jump in here...

So I did my first sync using onedrive -s. It finished in the terminal dialog and I closed the dialog. How does the bi-directional sync continue? Is it running under the covers? I think this is what I am missing.

**Reply by abraunegg** [reply]
When you perform a 'sync' using `--sync` or `-s` this is a one-time sync, and *is* by default bi-directional.

> How does the bi-directional sync continue? Is it running under the covers? I think this is what I am missing.

To do this, in a terminal window, you switch to using `--monitor` or `-m` and that performs the continous running until you use CTRL-C to exit that process.

To have the client running in the background using `--monitor` .. you need to configure a systemd service and run the client that way via the systemd service so that it automatically starts at system boot, or user login.

**Comment by stevefxp1** [comment]
Thanks for the quick response.

Ok that makes sense. How does the OneDrive Client for Linux GUI add to the equation, from a config perspective. 

For example I have installed your pkg and I have fixed the curl issues. I am ready to run the authorization via MS. After that I would normally just run your app via cli and setup the service for continuous operation.

Does the Onedrive Client for Linux GUI take care of getting your app running or do I still need to do the cli pieces? Is there a posting that talks to this?

**Reply by abraunegg** [reply]
@stevefxp1 
> Ok that makes sense. How does the OneDrive Client for Linux GUI add to the equation, from a config perspective.
> ...
> Does the Onedrive Client for Linux GUI take care of getting your app running or do I still need to do the cli pieces? Is there a posting that talks to this?

Due to how OneDriveGUI has been built and operates, it manually runs its own configuration / instance running `--monitor`. You will need to use the GUI to authenticate *that* specific instance. I believe you can import an existing configuration, but all of that you really should ask your OneDriveGUI questions here: https://github.com/bpozdena/OneDriveGUI/discussions
